clang/lib/Sema/SemaConcept.cpp

```diff
@@ -514,10 +509,87 @@ static ExprResult calculateConstraintSatisfaction(
   return SubstitutedAtomicExpr;
 }
 
-static UnsignedOrNone EvaluateFoldExpandedConstraintSize(
-    Sema &S, const FoldExpandedConstraint &FE, const NamedDecl *Template,
+static ExprResult calculateConstraintSatisfaction(
+    Sema &S, const AtomicConstraint &Constraint, const NamedDecl *Template,
     SourceLocation TemplateNameLoc, const MultiLevelTemplateArgumentList &MLTAL,
-    ConstraintSatisfaction &Satisfaction) {
+    ConstraintSatisfaction &Satisfaction,
+    UnsignedOrNone PackSubstitutionIndex) {
+
+  llvm::SmallVector<TemplateArgument> SubstitutedOuterMost;
+  std::optional<MultiLevelTemplateArgumentList> SubstitutedArgs =
+      SubstitutionInTemplateArguments(S, Constraint, Template, MLTAL,
+                                      SubstitutedOuterMost, Satisfaction,
+                                      PackSubstitutionIndex);
+  if (!SubstitutedArgs) {
+    Satisfaction.IsSatisfied = false;
+    return ExprEmpty();
+  }
+
+  unsigned Size = Satisfaction.Details.size();
+  llvm::SmallVector<TemplateArgument, 4> FlattenedArgs;
+  for (auto List : *SubstitutedArgs)
+    for (const TemplateArgument &Arg : List.Args)
+      FlattenedArgs.emplace_back(S.Context.getCanonicalTemplateArgument(Arg));
+
+  llvm::FoldingSetNodeID ID;
+  ID.AddPointer(Constraint.getConstraintExpr());
+  ID.AddInteger(FlattenedArgs.size());
+  for (auto &Arg : FlattenedArgs)
+    Arg.Profile(ID, S.Context);
+  ID.AddInteger(PackSubstitutionIndex.toInternalRepresentation());
+  unsigned CacheKeyHash = ID.ComputeHash();
```

Comment:
You could compute the hash without having a FlattenedArgs vector (and I don't think you need to hash the size)